File: c-friends/gradient_descent_solution.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#include "samples.h"


double w1, w0;
double learning_rate;

double hw(double x) {
	return w1 * x + w0;
}

double loss(double w1, double w0) {
	double sum = 0;
	for (int i = 0; i < SAMPLES_COUNT; i++) {
		sum += pow(hw(SAMPLES[i][0]) - SAMPLES[i][1], 2);
	}
	return sum;
}

double gradient_w0() {
	double sum = 0;
	for (int i = 0; i < SAMPLES_COUNT; i++) {
		sum += hw(SAMPLES[i][0]) - SAMPLES[i][1];
	}
	return learning_rate * sum;
}
double gradient_w1() {
	double sum = 0;
	for (int i = 0; i < SAMPLES_COUNT; i++) {
		sum += (hw(SAMPLES[i][0]) - SAMPLES[i][1]) * SAMPLES[i][0];
	}
	return learning_rate * sum;
}

void main() {
	srandom(time(NULL));
	w1 = ((double)rand() / RAND_MAX - 0.5) * 20, w0 = ((double)rand() / RAND_MAX - 0.5) * 200;
	learning_rate = 0.001;
	
	double g0 = INFINITY, g1 = INFINITY;
	for (int iteration = 0; pow(g0, 2) + pow(g1, 2) > 1e-6; iteration++) {
		printf("Iteration %i: w = (%.2f, %.2f), loss = %.2f, g = (%.2f, %.2f)\n", iteration, w0, w1, loss(w1, w0), g0, g1);
		
		g0 = gradient_w0(), g1 = gradient_w1();
		w0 -= g0, w1 -= g1;
	}
	
	printf("Samples:\n");
	printf("x, y, hw(x)\n");
	for (int i = 0; i < SAMPLES_COUNT; i++) {
		printf("%i, %i, %.2f\n", SAMPLES[i][0], SAMPLES[i][1], hw(SAMPLES[i][0]));
	}
}

File: gradient_descent_solution.py
#!/usr/bin/python3
import json
with open('samples.json') as samples_file:
	samples = json.load(samples_file)

import math
import random

def hw(x):
	return w1 * x + w0

def loss(samples, w1, w0):
	return sum([(hw(x) - y) ** 2 for [x, y] in samples])

def gradient_w0(samples):
	return sum([hw(x) - y for [x, y] in samples])
def gradient_w1(samples):
	return sum([(hw(x) - y) * x for [x, y] in samples])

w1, w0 = random.uniform(-10, 10), random.uniform(-100, 100)
learning_rate = 0.001

iteration = 0
g0 = g1 = math.inf
while g0 ** 2 + g1 ** 2 > 1e-6:
	iteration += 1
	print(f"Iteration {iteration}: w = ({w0:.2f}, {w1:.2f}), loss = {loss(samples, w1, w0):.2f}, g = ({g0:.2f}, {g1:.2f})")
	
	g0, g1 = gradient_w0(samples), gradient_w1(samples)
	w0 -= learning_rate * g0
	w1 -= learning_rate * g1

print("Samples:")
print("x, y, hw(x)")
for sample in samples:
	print(f"{sample[0]}, {sample[1]}, {hw(sample[0]):.2f}")
# extra line for satisfaction of answer to everything
